storage: introduce new backend localdisk in nydusd

adam3q opened the pull request:
The localdisk backend adds support for storing images in disks. In this scenario, each layer of the blob is stored in partitions, and multiple partitions are addressed in the local raw disk via the GUID partition table (GPT), which means that this disk stores the entire image.

Currently, generating a localdisk image through `nydusify` is not supported for the time being. You need to use the `nydus-localdisk` tool to complete this step.
Test document located at: https://github.com/adamqqqplay/nydus-localdisk/blob/master/README.md

This PR updates: #971 

Signed-off-by: Qinqi Qu <quqinqi@linux.alibaba.com>

changweige commented on storage/src/backend/localdisk.rs:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    // Disk names in GPT tables cannot store full 64-byte blob ids, so we should truncate them to 32 bytes.
+    fn truncate_blob_id(blob_id: &str) -> LocalDiskResult<&str> {
```
How about storing binary rather than strings? Then 32 bytes can hold the whole blob id

adamqqqplay replied:
> How about storing binary rather than strings? Then 32 bytes can hold the whole blob id

Very good suggestion, I will try to modify this.

adamqqqplay replied:
I tried to use the storage method of Hex to String, but there are still problems in the implementation, and I will improve it later.

In the future, I will use two different methods to obtain the blob_id in get_blob() to maintain compatibility with different versions of the localdisk image.

changweige commented:
How to create partitions filled with blobs and run with this PR?

adamqqqplay commented:
> How to create partitions filled with blobs and run with this PR?

https://github.com/adamqqqplay/nydus-localdisk
You could use this small tools to generate a localdisk image file and then mount it to the server via a loop device. I'll be porting this feature to nydusify shortly.

imeoer commented on storage/src/backend/localdisk.rs:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    // Disk names in GPT tables cannot store full 64-byte blob ids, so we should truncate them to 32 bytes.
+    fn truncate_blob_id(blob_id: &str) -> LocalDiskResult<&str> {
```
If the blob id is truncated in GPT table, how do we ensure no conflicts between different blobs with the same sha256 prefix?

adamqqqplay replied:
When the output of SHA-256 is truncated to 128 bits, its collision resistance is still 2^64, which is still enough to prevent blob collisions. And in the SHA standard, the result is also allowed to be truncated.

Please refer to:
https://security.stackexchange.com/questions/34796/truncating-the-output-of-sha256-to-128-bits
https://security.stackexchange.com/questions/72673/how-bad-is-it-to-truncate-a-hash/72685#72685

Despite this, I'll try to store the full blob id in the future, while maintaining compatibility with images generated in the past.

adamqqqplay replied:
> If the blob id is truncated in GPT table, how do we ensure no conflicts between different blobs with the same sha256 prefix?

Problem fixed in latest commit. Now the full 64-bit layer digest (blob id) stored in the Partition Name and Partition GUID respectively.

At the same time, forward and backward compatibility is maintained.

adamqqqplay commented on docs/nydusd.md:
````diff
@@ -151,6 +151,25 @@ We are working on enabling cloud-hypervisor support for nydus.
 ```
 
+##### Localdisk Backend
+
````
Ok, I'll add more description here.

imeoer commented on docs/nydusd.md:
````diff
@@ -151,6 +151,31 @@ We are working on enabling cloud-hypervisor support for nydus.
 }
 ```
 
+##### Localdisk Backend
````
Maybe mark this `Localdisk Backend (Experimental)` before we have supported it in nydusify.

adamqqqplay replied:
> Maybe mark this `Localdisk Backend (Experimental)` before we have supported it in nydusify.

OK, that's good.

adam3q committed:
storage: introduce new backend `localdisk`

This patch adds a new storage backend localdisk which supports reading
blobs on disk. In this scenario, each layer of the blob is stored in
partitions, and multiple partitions are addressed in the local raw
disk via the GUID partition table (GPT), which means that this disk
stores the entire image.

Signed-off-by: Qinqi Qu <quqinqi@linux.alibaba.com>
